### ISSUE TITLE
xwidget evil bindings are not being activated

### ISSUE THREAD

**Issue by pkazmier** [hlissner/doom-emacs#4080]
**What did you expect to happen?**
I expect the xwidget evil bindings defined in evil-collection to be activated upon entering xwidget-webkit-mode.

**What actually happened?**
The evil bindings are not activated when entering widget-webkit-mode.

**Additional details:**
I've validated this is not an issue with my local config as it occurs in the Doom Sandbox (`C-c C-p` version).

No errors are reported. Below are contents of `*Messages*` buffer:
<details><pre>
For information about GNU Emacs and the GNU system, type C-h C-a.
Loading /Users/xxxxxxx/.emacs.d/core/core.el (source)...done
DOOM Initializing Doom
Debug mode on
DOOM [editor/evil] Initialized evil-collection-explain-pause-mode (too early!)
DOOM [editor/evil] Initialized evil-collection-comint (too early!)
DOOM [editor/evil] Initialized evil-collection-custom (too early!)
DOOM [editor/evil] Initialized evil-collection-help (too early!)
DOOM [editor/evil] Initialized evil-collection-profiler (too early!)
Source file ‘/usr/local/share/emacs/site-lisp/mu/mu4e/mu4e-meta.el’ newer than byte-compiled file; using older file
DOOM Running doom hook: evil-mode
DOOM Running doom hook: +evil--init-popup-rules-h
DOOM Running doom hook: +workspaces-init-h
DOOM Running doom hook: doom-modeline-mode
DOOM Running doom hook: doom-init-fonts-h
DOOM Running doom hook: doom-init-leader-keys-h
DOOM Running doom hook: tramp-register-archive-file-name-handler
DOOM Running doom hook: collapse-delayed-warnings
DOOM Running doom hook: display-delayed-warnings
DOOM Running doom hook: doom-load-packages-incrementally-h
DOOM Running doom hook: doom-reset-file-handler-alist-h
DOOM Running doom hook: doom-modeline-refresh-font-width-cache
DOOM Running doom hook: #[0 \301\302\303\304\305\306\307\310\300!\311"\312$#\207 [(#0) run-at-time 10 nil make-byte-code 0 \301\302\300\242"\210\303 \207 vconcat vector [remove-hook window-setup-hook persp-update-frames-buffer-predicate] 3] 9]
DOOM Running doom hook: doom-init-ui-h
DOOM Running doom hook: +doom-dashboard-init-h
DOOM [ui/doom-dashboard] Reloading dashboard at 11:57:09
DOOM [ui/doom-dashboard] Changed dashboard’s PWD to ~/.emacs.d/
DOOM Running doom hook: doom-init-theme-h
DOOM Running doom hook: window-divider-mode
DOOM Running doom hook: +popup-mode
DOOM Running doom hook: (closure ((force-p) t) (&rest _) (add-hook 'hack-local-variables-hook #'doom-run-local-var-hooks-h) (add-hook 'after-change-major-mode-hook #'doom-run-local-var-hooks-maybe-h 'append) (add-hook 'doom-first-input-hook #'gcmh-mode) (let ((fn (intern (format %s-h 'doom-first-input-hook)))) (fset fn #'(lambda (&rest _) (run-hook-wrapped 'doom-first-input-hook #'doom-try-run-hook) (set 'doom-first-input-hook nil))) (put 'doom-first-input-hook 'permanent-local t) (let ((--dolist-tail-- (list 'pre-command-hook))) (while --dolist-tail-- (let ((on (car --dolist-tail--))) (if (functionp on) (advice-add on :before fn) (add-hook on fn)) (setq --dolist-tail-- (cdr --dolist-tail--)))))) (let ((fn (intern (format %s-h 'doom-first-file-hook)))) (fset fn #'(lambda (&rest _) (run-hook-wrapped 'doom-first-file-hook #'doom-try-run-hook) (set 'doom-first-file-hook nil))) (put 'doom-first-file-hook 'permanent-local t) (let ((--dolist-tail-- (list 'after-find-file 'dired-initial-position-hook))) (while --dolist-tail-- (let ((on (car --dolist-tail--))) (if (functionp on) (advice-add on :before fn) (add-hook on fn)) (setq --dolist-tail-- (cdr --dolist-tail--)))))) (let ((fn (intern (format %s-h 'doom-first-buffer-hook)))) (fset fn #'(lambda (&rest _) (run-hook-wrapped 'doom-first-buffer-hook #'doom-try-run-hook) (set 'doom-first-buffer-hook nil))) (put 'doom-first-buffer-hook 'permanent-local t) (let ((--dolist-tail-- (list 'after-find-file 'doom-switch-buffer-hook))) (while --dolist-tail-- (let ((on (car --dolist-tail--))) (if (functionp on) (advice-add on :before fn) (add-hook on fn)) (setq --dolist-tail-- (cdr --dolist-tail--)))))))
DOOM Running doom hook: doom-init-tty-h
DOOM Running doom hook: doom-display-benchmark-h
Doom loaded -1 packages across 45 modules in 2.238s
DOOM [editor/evil] Initialized evil-collection-compile 
DOOM Running doom hook: global-flycheck-mode
DOOM [editor/evil] Initialized evil-collection-flycheck 
DOOM [editor/evil] Initialized evil-collection-popup 
DOOM Running doom hook: undo-fu-mode
DOOM Running doom hook: ws-butler-global-mode
DOOM Running doom hook: smartparens-global-mode
DOOM Running doom hook: show-paren-mode
DOOM Running doom hook: winner-mode
DOOM [ui/doom-dashboard] Reloading dashboard at 11:57:09
DOOM [ui/doom-dashboard] Changed dashboard’s PWD to ~/.emacs.d/
DOOM Running doom hook: +doom-dashboard-init-h
DOOM Running doom hook: doom-init-theme-h
DOOM Running doom hook: window-divider-mode
DOOM Running doom hook: +popup-mode
Doom loaded -1 packages across 45 modules in 2.238s
Explain-pause-mode enabled.
DOOM Incrementally loading tree-widget [2 times]
DOOM Incrementally loading recentf [2 times]
DOOM Incrementally loading savehist
DOOM Incrementally loading yasnippet
DOOM [editor/evil] Initialized evil-collection-elisp-mode 
DOOM Incrementally loading with-editor
DOOM Incrementally loading yasnippet
DOOM Running doom hook: gcmh-mode
DOOM Running doom hook: evil-goggles-mode
DOOM Running doom hook: ivy-mode
DOOM [editor/evil] Initialized evil-collection-dired 
DOOM [editor/evil] Initialized evil-collection-xref 
DOOM [editor/evil] Initialized evil-collection-ivy 
DOOM Running doom hook: global-company-mode
DOOM Running doom hook: better-jumper-mode
DOOM Running doom hook: savehist-mode
DOOM Running doom hook: which-key-mode
DOOM [editor/evil] Initialized evil-collection-which-key 
DOOM [editor/evil] Loading deferred package evil-snipe from doom--after-call-evil-snipe-h
DOOM [editor/evil] Loading deferred package evil-escape from doom--after-call-evil-escape-h
DOOM [editor/evil] Loading deferred package evil-easymotion from doom--after-call-evil-easymotion-h
DOOM Loading deferred package server from doom--after-call-server-h
DOOM Incrementally loading with-editor
DOOM Incrementally loading git-commit
DOOM [editor/evil] Initialized evil-collection-epa 
DOOM [editor/evil] Initialized evil-collection-log-edit 
DOOM Incrementally loading package
DOOM Incrementally loading lv
DOOM Incrementally loading magit
DOOM [editor/evil] Initialized evil-collection-diff-mode 
DOOM [editor/evil] Initialized evil-collection-imenu 
DOOM [editor/evil] Initialized evil-collection-magit 
DOOM [editor/evil] Initialized evil-collection-calendar  [2 times]
DOOM [editor/evil] Initialized evil-collection-outline 
DOOM [editor/evil] Initialized evil-collection-grep 
DOOM [editor/evil] Initialized evil-collection-magit-todos 
DOOM Incrementally loading org-agenda
DOOM Incrementally loading org-capture
DOOM Incrementally loading org-agenda
DOOM Incrementally loading org-capture
DOOM [editor/evil] Initialized evil-collection-bookmark 
Garbage collecting...done (0.051s)
</pre></details>

**Steps to reproduce:**
1. Open Doom sandbox: `SPC h E` followed by `C-c C-p`
2. Open a web page via `M-x xwidget-webkit-browse-url`
3. Try to scroll up and down via `j`, `k`, `C-f`, or `C-b`

To workaround the problem, manually initialize the keybindings by evaluating this sexp: `(+evil-collection-init 'xwidget)`. After doing so, step three above will work as expected.

**System information:**
<details>

```
SYSTEM  type       darwin
        config     x86_64-apple-darwin19.6.0
        shell      /bin/zsh
        uname      Darwin 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64
        path       (~/.venv/awsrun/bin ~/bin ~/go/bin ~/.emacs.d/bin ~/.dotfiles/bin /usr/local/bin ~/.cache/zsh4humans/bin ~/.cache/zsh4humans/junegunn/fzf/bin /usr/bin /bin /usr/sbin /sbin /Library/TeX/texbin /usr/local/go/bin /usr/local/Cellar/emacs-plus@27/27.1/libexec/emacs/27.1/x86_64-apple-darwin19.6.0/)
EMACS   dir        ~/.emacs.d/
        version    27.1
        build      Oct 10, 2020
        buildopts  --disable-dependency-tracking --disable-silent-rules --enable-locallisppath=/usr/local/share/emacs/site-lisp --infodir=/usr/local/Cellar/emacs-plus@27/27.1/share/info/emacs --prefix=/usr/local/Cellar/emacs-plus@27/27.1 --with-xml2 --with-gnutls --without-dbus --with-imagemagick --with-modules --with-rsvg --with-xwidgets --with-ns --disable-ns-self-contained
        features   RSVG IMAGEMAGICK GLIB NOTIFY KQUEUE ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS MODULES THREADS XWIDGETS JSON PDUMPER LCMS2 GMP
        traits     (server-running envvar-file)
DOOM    dir        ~/.doom.d/
        version    2.0.9
        build      HEAD -> develop 3e2c98a0 2020-10-10 23:19:06 -0400
        elc-files  0
        modules    (:completion (company +childframe) ivy :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash ophints (popup +defaults) treemacs vc-gutter vi-tilde-fringe window-select workspaces zen :editor (evil +everywhere) file-templates fold (format +onsave) multiple-cursors snippets :emacs dired electric (ibuffer +icons) undo vc :term vterm :checkers syntax :tools (eval +overlay) lookup lsp magit :lang data emacs-lisp (go +lsp) markdown (org +dragndrop) (python +lsp) sh yaml :email mu4e :app calendar (rss +org) :config (default +bindings +smartparens))
        packages   (n/a)
        unpin      (doom-themes)
        elpa       (n/a)
```
</details>

**Comment by hlissner** [comment]
[We wait for xwidget-webkit-mode to be activated](https://github.com/hlissner/doom-emacs/blob/develop/modules/editor/evil/init.el#L280-L281) before we load evil-collection-xwidgets. Perhaps you can suggest a better place to hook it to?

**Comment by pkazmier** [comment]
I’m confused now. Doesn’t xwidget-webkit-mode get activated when I can call xwidget-webkit-browse-url? When the browser window opens, it `C-h m` says it is in xwidget-webkit mode, so why would the hook not kick in?

**Comment by pkazmier** [comment]
@hlissner I'm not too familiar with Elisp, but I tried to debug further. So, the root of this issue is that `(+evil-collection-init 'xwidget)` is never being called at any point in time. I would expect it to be called after I invoke `xwidget-webkit-browse-url`.  It never does because `+evil-collection-init` would have logged a message (below), so we can prove that it is never being invoked.
```
DOOM [editor/evil] Initialized evil-collection-xwidget 
```

As you point out, a transient hook used to lazily load these bindings. So, this is the part where it gets fuzzy for me. While poking around variables and functions with `C-h v` and `C-h f`, I noticed that `xwidget-webkit-mode` shows up as a _variable_ instead of as a function like some other modes. The value associated with variable is the transient function installed by `add-transient-hook`:

<img width="495" alt="image" src="https://user-images.githubusercontent.com/747855/95706800-8206ca00-0c1d-11eb-821b-11a9848b2ae3.png">

Is this normal? Should this be showing up as a function? Could this be the reason it is never invoked? Because when I call `xwidget-webkit-browse-url`, and it creates a webkit buffer, it ultimately has to invoke a function called `xwidget-webkit-mode` based on the `define-derived-mode` used to create the major mode, but this looks like a variable and not a function.

I also thought it was odd that _before_ I call `xwidget-webkit-browse-url` there is only a _variable_ called `xwidget-webkit-mode` as mentioned above. But after I open a web page, there is now a _function_ called `xwidget-webkit-mode` too! And, interestingly, the _variable_ version of the same name is still pointing to the transient hook. I.e. it was never called as it didn't remove itself, nor do we see a log message.

I'm way out of my league here, but I'm hoping you know the answer!
